### PR TITLE
Add support for adding or removing classes at scroll positions

### PR DESCRIPTION
**Pull request by LiamBigelow**
Something I needed for a current project was the ability to trigger CSS keyframe animations, and perform binary element changes (e.g. changing display, or some other non-animatable property). 

Rather than trigger these in JS, I wanted the scroll position logic to be specified alongside my lax animations.

- Added the data-lax-class:\<name\> syntax. 
  - When `value >= 1` we add the class
  - When `value <= 0` we remove it. 
  - Anything `0 > value < 1` means no change. 
This can be used to add a buffer in between adding the class and removing it.
- Added the original classnames to a map for each element on initialisation (similar to original styles) 
- Classes are tracked internally, as calling `classList.add()`/`classList.remove()` needlessly was a performance hog
- Added section to the README

Examples:
```
// Add a class only at the very bottom of the document
data-lax-class:bottom="(document.body.scrollHeight-vh-1) 0, (document.body.scrollHeight-vh) 1"

// Add a class only at the very top of the document
data-lax-class:top="0 1, 1 0"

// Add a class at the very bottom of the document, and remove it at the very top
data-lax-class:end="0 0, (document.body.scrollHeight-vh) 1"

// Add a class when the element is going off the top of the viewport
data-lax-class:departing="0 0, -1 1" data-lax-anchor="self"
```

---

Unrelated to this edit, but another fix:

I tweaked what is now [Line 358](https://github.com/alexfoxy/laxxx/compare/master...LiamBigelow:master#diff-37023a7fde89bd197246fd50c9566349R358) to check `anchorTop !== undefined`, rather than just being truth-ey. In the case where the anchor element was at the top of the document, its top offset was 0. That failed the `anchorTop ?` check, so it wouldn't be used as an anchor.

---

Let me know if you need any edits, or more clarification.

### CODE REVIEW

**Comment by alexfoxy** [comment]
Nice! Looks cool - do you have a demo anywhere ?

**Comment by willryanuk** [comment]
This is also something that I'd really like to be incorporated into the library. Pinning animation to scroll can sometimes mean fades/zooms etc happen too quickly if the user is scrolling super quick, so would prefer to trigger the animations via classes, with constant animation times.

**Comment by LiamBigelow** [comment]
I’ll have a chance to get a demo up next week.

**Comment by willryanuk** [comment]
Yeah, I'd love this feature. It's the missing part of the puzzle for me. Otherwise, this library is really well put together, simple yet very powerful :)

File: lib/lax.js [hunk 1]
```diff
@@ -479,12 +493,14 @@ function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
           anchorTop = o.anchorTop,
           transforms = o.transforms,
           spriteData = o.spriteData,
-          el = o.el;
-      var r = anchorTop ? anchorTop - lastY : lastY;
+          el = o.el,
+          classes = o.classes;
+      var r = anchorTop !== undefined ? anchorTop - lastY : lastY;
```
Comment (mtskf) [inline lib/lax.js]: Changing this line causes error.

File: src/lax.js [hunk 2]
```diff
@@ -339,15 +353,17 @@
     }
 
     lax.updateElement = (o) => {
-      const { originalStyle, anchorTop, transforms, spriteData, el } = o
+      const { originalStyle, anchorTop, transforms, spriteData, el, classes } = o
 
-      let r = anchorTop ? anchorTop-lastY : lastY
+      let r = anchorTop !== undefined ? anchorTop-lastY : lastY
```
Comment (mtskf) [inline src/lax.js]: Changing this line causes error.

**Comment by alexfoxy** [comment]
@LiamBigelow did you ever fix the errors @mtskf pointed out?

**Comment by PubliAlex** [comment]
Any news about it ? It's closed cause it has been committed or still not possible with laxxx ?

**Comment by bglw** [comment]
This was closed as I never picked it back up (apologies) — lax has done a major upgrade since this PR so I don't imagine the code here is relevant any more.

**Pull request closed without merging**